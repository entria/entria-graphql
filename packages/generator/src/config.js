/* @flow */
import path from 'path';
import merge from 'lodash.merge';
import fs from 'fs';
import pkgDir from 'pkg-dir';

const rootPath = pkgDir.sync('.') || '.';

let cacheConfig = null;

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath: string) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === DIRECTORY_TYPE.SOURCE) {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
        [DIRECTORY_TYPE.SRC]: config.directories.source,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories.module}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

export const DIRECTORY_TYPE = {
  SOURCE: 'source',
  SRC: 'src',
  MODULE: 'module',
  CONNECTION: 'connection',
  LOADER: 'loader',
  MODEL: 'model',
  MUTATION: 'mutation',
  MUTATION_TEST: 'mutation_test',
  TYPE: 'type',
  TYPE_TEST: 'type_test',
};

type DirectoryType = $Values<typeof DIRECTORY_TYPE>;
type Directories = {
  source: string,
  module: string,
  connection: string,
  loader: string,
  model: string,
  mutation: string,
  mutation_test: string,
  type: string,
  type_test: string,
  interface: string,
};

type Files = {
  schema: string,
};

type Config = {
  directories: Directories,
  files: Files,
}
/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = (): Config => {
  // if (cacheConfig) return cacheConfig;

  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const config = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    merge(config, customConfig);

    cacheConfig = config;

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return config;
  } catch (err) {
    cacheConfig = config;
    // Return the default config if the custom doesn't exist
    return config;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory: DirectoryType) => getCreateGraphQLConfig().directories[directory];
