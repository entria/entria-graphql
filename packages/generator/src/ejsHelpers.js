/* @flow */
/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = (text: string) => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text: string) =>
  text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
