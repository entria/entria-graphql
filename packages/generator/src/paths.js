/* @flow */
import path from 'path';

export const getModulePath = (source: string, moduleName: string) => {
  return path.join(source, moduleName);
};

export const getTestPath = (source: string) => {
  return path.join(source, '__tests__');
};

export const getMutationPath = (source: string) => {
  return path.join(source, 'mutation');
};
