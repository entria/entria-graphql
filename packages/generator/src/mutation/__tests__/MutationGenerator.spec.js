import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';
import fs from 'fs-extra';

import {
  getFileContent,
  getFixturePath,
} from '../../../test/helpers';

import { getConfigDir } from '../../config';
import { uppercaseFirstLetter } from '../../ejsHelpers';
import { getModulePath, getTestPath, getMutationPath } from '../../paths';

const mutationGenerator = path.join(__dirname, '..');

it('generate mutation files', async () => {
  const moduleName = 'example';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(mutationGenerator)
    .withArguments(name)
    .toPromise();

  const destinationDir = getConfigDir('mutation');

  const modulePath = getModulePath(destinationDir, moduleName);
  const mutationPath = getMutationPath(modulePath);
  const mutationTestPath = getTestPath(mutationPath);

  const mutationAddFilepath = path.join(mutationPath, `${name}AddMutation.js`);
  const mutationEditFilepath = path.join(mutationPath, `${name}EditMutation.js`);
  const mutationAddTestFilepath = path.join(mutationTestPath, `${name}AddMutation.spec.js`);
  const mutationEditTestFilepath = path.join(mutationTestPath, `${name}EditMutation.spec.js`);

  assert.file([
    mutationAddFilepath,
    mutationEditFilepath,
    mutationAddTestFilepath,
    mutationEditTestFilepath,
  ]);

  const files = {
    add: getFileContent(path.join(folder, mutationAddFilepath)),
    edit: getFileContent(path.join(folder, mutationEditFilepath)),
    addTest: getFileContent(path.join(folder, mutationAddTestFilepath)),
    editTest: getFileContent(path.join(folder, mutationEditTestFilepath)),
  };

  expect(files).toMatchSnapshot();
});

it('generate mutation files with schema', async () => {
  const moduleName = 'post';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(mutationGenerator)
    .inTmpDir(dir =>
      fs.copySync(
        getFixturePath(name),
        path.join(dir, 'src/modules/post/PostModel.js'),
      ),
    )
    .withArguments('Post Post')
    .toPromise();

  const destinationDir = getConfigDir('mutation');

  const modulePath = getModulePath(destinationDir, moduleName);
  const mutationPath = getMutationPath(modulePath);
  const mutationTestPath = getTestPath(mutationPath);

  const mutationAddFilepath = path.join(mutationPath, `${name}AddMutation.js`);
  const mutationEditFilepath = path.join(mutationPath, `${name}EditMutation.js`);
  const mutationAddTestFilepath = path.join(mutationTestPath, `${name}AddMutation.spec.js`);
  const mutationEditTestFilepath = path.join(mutationTestPath, `${name}EditMutation.spec.js`);

  assert.file([
    mutationAddFilepath,
    mutationEditFilepath,
    mutationAddTestFilepath,
    mutationEditTestFilepath,
  ]);

  const files = {
    add: getFileContent(path.join(folder, mutationAddFilepath)),
    edit: getFileContent(path.join(folder, mutationEditFilepath)),
    addTest: getFileContent(path.join(folder, mutationAddTestFilepath)),
    editTest: getFileContent(path.join(folder, mutationEditTestFilepath)),
  };

  expect(files).toMatchSnapshot();
});
