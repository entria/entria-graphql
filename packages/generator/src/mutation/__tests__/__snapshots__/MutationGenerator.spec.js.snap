// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate mutation files 1`] = `
Object {
  "add": "// @flow
import {
  GraphQLString,
  GraphQLNonNull,
} from 'graphql';
import {
  mutationWithClientMutationId,
  toGlobalId,
} from 'graphql-relay';

import ExampleModel from '../ExampleModel';

import ExampleLoader from '../ExampleLoader';
import ExampleConnection from '../ExampleConnection';

export default mutationWithClientMutationId({
  name: 'ExampleAdd',
  inputFields: {
    example: {
      type: new GraphQLNonNull(GraphQLString),
      description: 'My example field',
    },
  },
  mutateAndGetPayload: async ({ example }, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error('Unauthorized user');
    }

    // TODO: mutation logic

    return {
      // id: id, // ID of the newly created row
      error: null,
    };
  },
  outputFields: {
    exampleEdge: {
      type: ExampleConnection.edgeType,
      resolve: async ({ id }, args, context) => {
        // Load new edge from loader
        const example = await ExampleLoader.load(
          context, id,
        );

        // Returns null if no node was loaded
        if (!example) {
          return null;
        }

        return {
          cursor: toGlobalId('Example', example),
          node: example,
        };
      },
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "addTest": "import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../test/helper';

import { User } from '.';
import Example from './Example';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should not allow anonymous user', async () => {
  //language=GraphQL
  const query = \`
    mutation M($example: String) {
      ExampleAdd(input: {
        example: $example
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext();

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});

it('should create a record on database', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
  });

  await user.save();

  //language=GraphQL
  const query = \`
    mutation M($example: String) {
      ExampleAdd(input: {
        example: $example
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});
",
  "edit": "// @flow
import {
  GraphQLID,
  GraphQLString,
  GraphQLNonNull,
} from 'graphql';
import {
  mutationWithClientMutationId,
  fromGlobalId,
} from 'graphql-relay';

import ExampleModel from '../ExampleModel';

import ExampleType from '../ExampleType';
import ExampleLoader from './ExampleLoader';

export default mutationWithClientMutationId({
  name: 'ExampleEdit',
  inputFields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
    example: {
      type: GraphQLString,
    },
  },
  mutateAndGetPayload: async (args, context) => {
    const { user } = context;

    // Verify if user is authorized
    if (!user) {
      throw new Error('Unauthorized user');
    }

    const {
      id,
      example,
    } = args;

    // Check if the provided ID is valid
    const example = await ExampleModel.findOne({
      _id: fromGlobalId(id).id,
    });

    // If not, throw an error
    if (!example) {
      throw new Error('Invalid exampleId');
    }

    // TODO: mutation logic

    // Clear dataloader cache
    ExampleLoader.clearCache(context, example._id);

    return {
      id: example._id,
      error: null,
    };
  },
  outputFields: {
    example: {
      type: ExampleType,
      resolve: (obj, args, context) => ExampleLoader.load(context, obj.id),
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "editTest": "import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../test/helper';

import User from './User';
import Example from './Example';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should not allow anonymous user', async () => {
  //language=GraphQL
  const query = \`
    mutation M {
      ExampleEdit(input: {
        id: \\"Example Id\\"
        exampleField: \\"Example field\\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext();

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});

it('should create a record on database', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
  });

  await user.save();

  //language=GraphQL
  const query = \`
    mutation M {
      ExampleEdit(input: {
        id: \\"Example Id\\"
        exampleField: \\"Example field\\"
      }) {
        exampleFieldToRetrieve
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});
",
}
`;

exports[`generate mutation files with schema 1`] = `
Object {
  "add": "// @flow
import {
  GraphQLString,
  GraphQLID,
  GraphQLList,
  GraphQLNonNull,
} from 'graphql';
import {
  mutationWithClientMutationId,
  toGlobalId,
} from 'graphql-relay';

import PostModel from '../PostModel';

import PostLoader from '../PostLoader';
import PostConnection from '../PostConnection';

export default mutationWithClientMutationId({
  name: 'PostAdd',
  inputFields: {
    title: {
      type: GraphQLNonNull(GraphQLString),
    },
    user: {
      type: GraphQLNonNull(GraphQLID),
    },
    slug: {
      type: GraphQLString,
    },
    tags: {
      type: GraphQLList(GraphQLString),
    },
    oldSlugs: {
      type: GraphQLList(GraphQLString),
    },
    comments: {
      type: GraphQLList(GraphQLID),
    },
    externalComments: {
      type: GraphQLList(GraphQLID),
    },
  },
  mutateAndGetPayload: async (args, { user }) => {
    // Verify if user is authorized
    if (!user) {
      throw new Error('Unauthorized user');
    }

    const {
      title,
      user,
      slug,
      tags,
      oldSlugs,
      comments,
      externalComments,
    } = args;

    // Create new record
    const post = await new PostModel({
      title,
      user,
      slug,
      tags,
      oldSlugs,
      comments,
      externalComments,
    }).save();

    // TODO: mutation logic

    return {
      id: post._id,
      error: null,
    };
  },
  outputFields: {
    postEdge: {
      type: PostConnection.edgeType,
      resolve: async ({ id }, args, context) => {
        // Load new edge from loader
        const post = await PostLoader.load(
          context, id,
        );

        // Returns null if no node was loaded
        if (!post) {
          return null;
        }

        return {
          cursor: toGlobalId('Post', post),
          node: post,
        };
      },
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "addTest": "import { graphql } from 'graphql';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../test/helper';

import User from './User';
import Post from './Post';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should not allow anonymous user', async () => {
  //language=GraphQL
  const query = \`
    mutation M {
      PostAdd(input: {
        title: \\"Example value\\"
        user: \\"Example value\\"
        slug: \\"Example value\\"
        tags: \\"Example value\\"
        oldSlugs: \\"Example value\\"
        comments: \\"Example value\\"
        externalComments: \\"Example value\\"
      }) {
        postEdge {
          node {
            title
            user
            slug
            tags
            oldSlugs
            comments
            externalComments
          }
        }
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext();

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});

it('should create a record on database', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
  });

  await user.save();

  //language=GraphQL
  const query = \`
    mutation M {
      PostAdd(input: {
        title: \\"Example value\\"
        user: \\"Example value\\"
        slug: \\"Example value\\"
        tags: \\"Example value\\"
        oldSlugs: \\"Example value\\"
        comments: \\"Example value\\"
        externalComments: \\"Example value\\"
      }) {
        postEdge {
          node {
            title
            user
            slug
            tags
            oldSlugs
            comments
            externalComments
          }
        }
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});
",
  "edit": "// @flow
import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLID,
  GraphQLList,
} from 'graphql';
import {
  mutationWithClientMutationId,
  fromGlobalId,
} from 'graphql-relay';

import PostModel from '../PostModel';

import PostType from '../PostType';
import PostLoader from '../PostLoader';

export default mutationWithClientMutationId({
  name: 'PostEdit',
  inputFields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
    },
    title: {
      type: GraphQLNonNull(GraphQLString),
    },
    user: {
      type: GraphQLNonNull(GraphQLID),
    },
    slug: {
      type: GraphQLString,
    },
    tags: {
      type: GraphQLList(GraphQLString),
    },
    oldSlugs: {
      type: GraphQLList(GraphQLString),
    },
    comments: {
      type: GraphQLList(GraphQLID),
    },
    externalComments: {
      type: GraphQLList(GraphQLID),
    },
  },
  mutateAndGetPayload: async (args, context) => {
    const { user } = context;

    // Verify if user is authorized
    if (!user) {
      throw new Error('Unauthorized user');
    }

    const {
      id,
      title,
      user,
      slug,
      tags,
      oldSlugs,
      comments,
      externalComments,
    } = args;

    // Check if the provided ID is valid
    const post = await PostModel.findOne({
      _id: fromGlobalId(id).id,
    });

    // If not, throw an error
    if (!post) {
      throw new Error('Invalid postId');
    }

    // Edit record
    await post.update({
      title,
      user,
      slug,
      tags,
      oldSlugs,
      comments,
      externalComments,
    });

    // TODO: mutation logic

    // Clear dataloader cache
    PostLoader.clearCache(context, post._id);

    return {
      id: post._id,
      error: null,
    };
  },
  outputFields: {
    post: {
      type: PostType,
      resolve: (obj, args, context) => PostLoader.load(context, obj.id),
    },
    error: {
      type: GraphQLString,
      resolve: ({ error }) => error,
    },
  },
});
",
  "editTest": "import { graphql } from 'graphql';
import { toGlobalId } from 'graphql-relay';
import { schema } from '../../schema';
import {
  getContext,
  connectMongoose,
  clearDbAndRestartCounters,
  disconnectMongoose,
} from '../../../test/helper';

import User from './User';
import Post from './Post';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

it('should not allow anonymous user', async () => {
  // TODO: specify fields to create a new Post
  const post = new Post({
    title: 'Example value',
    user: 'Example value',
    slug: 'Example value',
    tags: 'Example value',
    oldSlugs: 'Example value',
    comments: 'Example value',
    externalComments: 'Example value',
  });

  await post.save();

  const postId = toGlobalId('Post', post._id);

  //language=GraphQL
  const query = \`
    mutation M {
      PostEdit(input: {
        id: \\"\${postId}\\"
        example: \\"Example Field to Update\\"
      }) {
        post {
          title
          user
          slug
          tags
          oldSlugs
          comments
          externalComments
        }
      }
    }
  \`;

  const variables = {};
  const rootValue = {};
  const context = getContext();

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});

it('should edit a record on database', async () => {
  const user = new User({
    name: 'user',
    email: 'user@example.com',
  });

  await user.save();

  // TODO: specify fields to create a new Post
  const post = new Post({
    title: 'Example value',
    user: 'Example value',
    slug: 'Example value',
    tags: 'Example value',
    oldSlugs: 'Example value',
    comments: 'Example value',
    externalComments: 'Example value',
  });

  await post.save();

  const postId = toGlobalId('Post', post._id);

  //language=GraphQL
  const query = \`
    mutation M {
      PostEdit(input: {
        id: \\"\${postId}\\"
        example: \\"Example Field to Update\\"
      }) {
        post {
          title
          user
          slug
          tags
          oldSlugs
          comments
          externalComments
        }
      }
    }
  \`;

  const variables = {
  };
  const rootValue = {};
  const context = getContext({ user });

  const result = await graphql(schema, query, rootValue, context, variables);

  expect(result).toMatchSnapshot();
});
",
}
`;
